okhttp/src/main/kotlin/okhttp3/internal/connection/RealConnectionPool.kt

```diff
@@ -286,7 +303,82 @@ class RealConnectionPool(
     return references.size
   }
 
+  fun setPolicy(
+    address: Address,
+    policy: ConnectionPool.AddressPolicy,
+  ) {
+    // TODO: it's quite awkward that we have a new policy object that doesn't contain the max
+    require(policy.minimumConcurrentCalls <= maxIdleConnections) {
+      "minimumConcurrentCalls must be <= maxIdleConnections"
+    }
+
+    // Race conditions are fine; worst case, we check for min connections more often than needed
+    if (!policies.contains(address)) {
+      minConnectionQueue.schedule(minConnectionTask(address))
+    }
+
+    policies[address] = policy
+  }
+
+  fun ensureMinimumConnections(address: Address): Long {
+    val policy = policies[address]!!
+    val relevantConnections = connections.filter { it.isEligible(address, null) }
+
+    when {
+      // We already have an existing http/2 connection that can handle all of our streams
+      relevantConnections.any { it.isMultiplexed } -> {}
+
+      // We already have enough connections to handle every call
+      relevantConnections.size >= policy.minimumConcurrentCalls -> {}
+
+      // This policy is unsatisfied -- open a connection!
+      relevantConnections.isEmpty() -> {
+        try {
+          val exchangeFinder = buildExchangeFinder(address)
+          val connection = exchangeFinder.find()
+          put(connection)
+
+          return 0 // run again immediately to create more connections if needed
+        } catch (ex: Exception) {
+          // TODO do something with exception?
```

Comment:
I couldn't find any prior art for what to do with exceptions that happen in some background thread like this. Do we log it somehow?

Reply:
I think generally we catch IOExceptions and handle. 

Unexpected runtime exceptions are allowed to crash the VM or whatever.

Known runtime exceptions in our dependencies, even bugs, are caught and handled.